File: 0001_two_sum/0001_two_sum.ts
/*
  Given an array of integers,
  return indices of the two numbers such that they add up to a specific target.

  You may assume that each input would have exactly one solution,
  and you may not use the same element twice.
*/

export const twoSum = (nums: number[], target: number) => {
  const hash: { [key: number]: number } = {};

  for (let index = 0; index < nums.length; index++) {
    const num = nums[index];

    if (hash.hasOwnProperty(num)) {
      return [hash[num], index];
    }

    hash[target - num] = index;
  }
};

File: cpp/0001_two_sum/0001_two_sum.cc
#include "0001_two_sum.h"
#include <iostream>
#include <unordered_map>
#include <vector>

using std::unordered_map;
using std::vector;

vector<int> Solution::twoSum(vector<int> &nums, int target) {
  unordered_map<int, int> umap;
  for (int i = 0; i < nums.size(); i++) {
    const int current = nums[i];

    if (umap.count(current) > 0) {
      return {i, umap[current]};
    }

    const int diff = target - nums[i];
    umap[diff] = i;
  }

  return {};
}
